pimd/pim_msg.c

```diff
@@ -296,3 +296,56 @@ size_t pim_msg_build_jp_groups(struct pim_jp_groups *grp,
 
 	return size;
 }
+#if PIM_IPV == 4
+uint8_t *pim_get_pim_protocol_from_iphdr(uint8_t *buf, size_t buf_size,
+					 uint32_t *pim_msg_len)
+{
+	struct ip *ip_hdr = (struct ip *)buf;
+	uint8_t *pim_msg;
+	size_t ip_hlen;
+
+	ip_hlen = ip_hdr->ip_hl << 2;
+	pim_msg = buf + ip_hlen;
+	*pim_msg_len = buf_size - ip_hlen;
+
+	return pim_msg;
+}
+#else
+uint8_t *pim_get_pim_protocol_from_iphdr(uint8_t *buf, size_t buf_size,
+					 uint32_t *pim_msg_len)
+{
+	const struct ip6_hdr *ip6_hdr;
+	uint8_t next_hdr;
+	uint32_t parsed_len = 0;
+	uint32_t ext_hdr_len = 0;
+
+	if (buf_size < sizeof(struct ip6_hdr))
+		return NULL;
+
+	ip6_hdr = (const struct ip6_hdr *)buf;
+
+	if (!ip6_hdr->ip6_nxt) {
+		return NULL;
+	}
+
+	next_hdr = ip6_hdr->ip6_nxt;
+	ext_hdr_len = buf_size - sizeof(struct ip6_hdr);
+	while (buf_size - parsed_len >= 8) {
+		switch (next_hdr) {
+		case PIM_IP_PROTO_PIM:
+			/* The protocol is found */
+			*pim_msg_len = ext_hdr_len;
+			return buf;
+		default:
```

Comment:
This isn't how ipv6 extension headers work

Only some values are extension headers (0, 43, 51, 60) and the length calculation is different for 51.

(You can ignore 44, fragmentation should be handled by the kernel & we won't try to reassemble fragments)

Reply:
This api is basically not needed any more. Since the kernel is stripping off the ipv6 headers. I will remove this commit.